enterprise/server/remote_execution/container/container.go

```diff
@@ -401,69 +410,157 @@ func GetPullCredentials(env environment.Env, props *platform.Properties) (PullCr
 	return PullCredentials{}, nil
 }
 
-// TracedCommandContainer is a wrapper that creates tracing spans for all CommandContainer methods.
+// TracedCommandContainer is a wrapper that creates tracing spans for all
+// CommandContainer methods. It also provides some basic protection against race
+// conditions, such as preventing Remove() from being called twice or while
+// other operations are in progress.
 type TracedCommandContainer struct {
-	Delegate CommandContainer
 	implAttr attribute.KeyValue
+
+	// This mutex is used only to prevent Remove() from being called
+	// concurrently with other operations. Exec() may be called concurrently
+	// with Pause() and Unpause() because persistent workers are implemented
+	// using a long-running Exec() operation.
+	mu       sync.RWMutex
+	removed  bool
+	Delegate CommandContainer
 }
 
 func (t *TracedCommandContainer) Run(ctx context.Context, command *repb.Command, workingDir string, creds PullCredentials) *interfaces.CommandResult {
 	ctx, span := tracing.StartSpan(ctx, trace.WithAttributes(t.implAttr))
 	defer span.End()
+
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.removed {
+		return &interfaces.CommandResult{ExitCode: noExitCode, Error: ErrRemoved}
+	}
+
 	return t.Delegate.Run(ctx, command, workingDir, creds)
 }
 
 func (t *TracedCommandContainer) IsImageCached(ctx context.Context) (bool, error) {
 	ctx, span := tracing.StartSpan(ctx, trace.WithAttributes(t.implAttr))
 	defer span.End()
+
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.removed {
+		return false, ErrRemoved
+	}
+
 	return t.Delegate.IsImageCached(ctx)
 }
 
 func (t *TracedCommandContainer) PullImage(ctx context.Context, creds PullCredentials) error {
 	ctx, span := tracing.StartSpan(ctx, trace.WithAttributes(t.implAttr))
 	defer span.End()
+
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.removed {
+		return ErrRemoved
+	}
+
 	return t.Delegate.PullImage(ctx, creds)
 }
 
 func (t *TracedCommandContainer) Create(ctx context.Context, workingDir string) error {
 	ctx, span := tracing.StartSpan(ctx, trace.WithAttributes(t.implAttr))
 	defer span.End()
+
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.removed {
+		return ErrRemoved
+	}
+
 	return t.Delegate.Create(ctx, workingDir)
 }
 
 func (t *TracedCommandContainer) Exec(ctx context.Context, command *repb.Command, opts *Stdio) *interfaces.CommandResult {
 	ctx, span := tracing.StartSpan(ctx, trace.WithAttributes(t.implAttr))
 	defer span.End()
+
+	log.Infof("Exec()...")
```

Comment:
Did you intend to leave in these debug logs?

Reply:
woops, thanks for catching this!